File: sdk/postgres.go
package ocel

import (
	"context"
	"crypto/x509"
	"errors"
	"fmt"
	"net"
	"net/url"
	"runtime"
	"strconv"
	"sync"

	"github.com/jackc/pgx/v5/pgxpool"

	resourcesv1 "ocel.dev/internal/proto/app/resources/v1"
	bindingsv1 "ocel.dev/internal/proto/common/bindings/v1"
)

const defaultPostgresVersion = "17"

// A PostgresOption tunes the database [Postgres] declares.
type PostgresOption func(*resourcesv1.PostgresConfig)

// PostgresVersion declares the major postgres version to provision, "17" by default.
func PostgresVersion(v string) PostgresOption {
	return func(c *resourcesv1.PostgresConfig) { c.Version = v }
}

// A PostgresDB is a postgres database an app declares and reads its binding from.
type PostgresDB struct {
	name string

	once sync.Once
	pool *pgxpool.Pool
	err  error
}

// Postgres declares a postgres database named name and returns the handle an app
// reads it through. Call it from a file under the project's discovery folder: during
// discovery the call is the declaration, and at runtime it reads the binding the
// deploy delivered for that name.
func Postgres(name string, opts ...PostgresOption) *PostgresDB {
	if discovering() {
		config := &resourcesv1.PostgresConfig{Version: defaultPostgresVersion}
		for _, opt := range opts {
			opt(config)
		}
		_, file, line, _ := runtime.Caller(1)
		err := declare(&resourcesv1.DeclareRequest{
			Resource: &resourcesv1.ResourceIdentifier{
				Type: resourcesv1.ResourceType_RESOURCE_TYPE_POSTGRES,
				Name: name,
			},
			Config: &resourcesv1.DeclareRequest_Postgres{Postgres: config},
			Source: fmt.Sprintf("%s:%d", file, line),
		})
		if err != nil {
			panic(fmt.Sprintf("ocel: declare postgres %q: %v", name, err))
		}
	}
	return &PostgresDB{name: name}
}

// Name is the name the database was declared under.
func (p *PostgresDB) Name() string { return p.name }

// ConnectionString is the postgres URL of the delivered binding: the record's url
// verbatim when it carries one, and otherwise one built from its host, port,
// database and credentials, percent-encoded, with its tls mode as sslmode. It
// fails when no binding was delivered for the name, and during discovery.
func (p *PostgresDB) ConnectionString() (string, error) {
	properties, err := p.properties("ConnectionString")
	if err != nil {
		return "", err
	}
	return connectionString(properties), nil
}

func connectionString(properties *bindingsv1.PostgresProperties) string {
	if properties.GetUrl() != "" {
		return properties.GetUrl()
	}
	query := url.Values{}
	if properties.GetTlsMode() != "" {
		query.Set("sslmode", properties.GetTlsMode())
	}
	dsn := url.URL{
		Scheme:   "postgres",
		User:     url.UserPassword(properties.GetUsername(), properties.GetPassword()),
		Host:     net.JoinHostPort(properties.GetHost(), strconv.Itoa(int(properties.GetPort()))),
		Path:     "/" + properties.GetDatabase(),
		RawQuery: query.Encode(),
	}
	return dsn.String()
}

// Pool is the connection pool over the delivered binding. A record under
// verify-full that names a CA trusts that CA for the server's certificate. The pool
// is opened on the first call and the same pool is returned on every one after. It
// fails when no binding was delivered for the name, and during discovery.
func (p *PostgresDB) Pool(ctx context.Context) (*pgxpool.Pool, error) {
	if discovering() {
		return nil, &UnprovisionedError{Resource: p.resource(), Access: "Pool"}
	}
	p.once.Do(func() {
		properties, err := p.properties("Pool")
		if err != nil {
			p.err = err
			return
		}
		config, err := pgxpool.ParseConfig(connectionString(properties))
		if err != nil {
			p.err = err
			return
		}
		if err := trustCA(config, properties.GetTlsCa()); err != nil {
			p.err = err
			return
		}
		p.pool, p.err = pgxpool.NewWithConfig(ctx, config)
	})
	return p.pool, p.err
}

func trustCA(config *pgxpool.Config, ca string) error {
	if ca == "" || config.ConnConfig.TLSConfig == nil {
		return nil
	}
	roots := x509.NewCertPool()
	if !roots.AppendCertsFromPEM([]byte(ca)) {
		return errors.New("ocel: the postgres binding's CA holds no PEM certificate")
	}
	config.ConnConfig.TLSConfig.RootCAs = roots
	for _, fallback := range config.ConnConfig.Fallbacks {
		if fallback.TLSConfig != nil {
			fallback.TLSConfig.RootCAs = roots
		}
	}
	return nil
}

func (p *PostgresDB) properties(access string) (*bindingsv1.PostgresProperties, error) {
	if discovering() {
		return nil, &UnprovisionedError{Resource: p.resource(), Access: access}
	}
	delivered, err := binding(p.name, bindingsv1.BindingType_BINDING_TYPE_POSTGRES)
	if err != nil {
		return nil, err
	}
	return delivered.GetPostgres(), nil
}

func (p *PostgresDB) resource() string {
	return fmt.Sprintf("postgres(%q)", p.name)
}

File: crates/ocel-sdk/src/postgres.rs
use crate::binding::{encoded, postgres};
use crate::declare::discovering;
use crate::proto::common::bindings::v1::PostgresProperties;
use crate::Error;

pub(crate) const KIND: &str = "postgres";

/// A postgres database an app declares and reads its binding from. A field of this type in a
/// struct deriving [`Resources`](macro@crate::Resources) is the declaration.
#[derive(Clone)]
pub struct Postgres {
    name: String,
    #[cfg(feature = "postgres")]
    pool: std::sync::Arc<tokio::sync::OnceCell<sqlx::PgPool>>,
}

impl Postgres {
    /// Take the handle for the database named `name`. Prefer
    /// [`Resources`](macro@crate::Resources), which declares the database as well as
    /// handing back its handle.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            #[cfg(feature = "postgres")]
            pool: std::sync::Arc::default(),
        }
    }

    /// The name the database was declared under, and the name its binding is delivered as.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// The postgres URL of the delivered binding: the record's url verbatim when it carries
    /// one, and otherwise one built from its host, port, database and credentials,
    /// percent-encoded, with its tls mode as `sslmode`. It fails when no binding was delivered
    /// for the name, and during discovery.
    pub fn connection_string(&self) -> Result<String, Error> {
        Ok(connection_string(&self.properties("connection_string")?))
    }

    /// The sqlx pool over the delivered binding, opened on the first call and returned as it
    /// stands on every one after. A record under verify-full that names a CA trusts that CA
    /// for the server's certificate. It fails when no binding was delivered for the name, and
    /// during discovery.
    #[cfg(feature = "postgres")]
    pub async fn pool(&self) -> Result<&sqlx::PgPool, Error> {
        if discovering() {
            return Err(self.unprovisioned("pool"));
        }
        self.pool
            .get_or_try_init(|| async {
                let options = connect_options(&self.properties("pool")?)?;
                Ok(sqlx::PgPool::connect_with(options).await?)
            })
            .await
    }

    fn properties(&self, access: &str) -> Result<PostgresProperties, Error> {
        if discovering() {
            return Err(self.unprovisioned(access));
        }
        postgres(&self.name)
    }

    fn unprovisioned(&self, access: &str) -> Error {
        Error::Unprovisioned {
            resource: format!("postgres(\"{}\")", self.name),
            access: access.to_string(),
        }
    }
}

fn connection_string(properties: &PostgresProperties) -> String {
    if !properties.url.is_empty() {
        return properties.url.clone();
    }
    let sslmode = if properties.tls_mode.is_empty() {
        String::new()
    } else {
        format!("?sslmode={}", encoded(&properties.tls_mode))
    };
    format!(
        "postgres://{}:{}@{}:{}/{}{}",
        encoded(&properties.username),
        encoded(&properties.password),
        properties.host,
        properties.port,
        encoded(&properties.database),
        sslmode,
    )
}

#[cfg(feature = "postgres")]
fn connect_options(
    properties: &PostgresProperties,
) -> Result<sqlx::postgres::PgConnectOptions, Error> {
    let options: sqlx::postgres::PgConnectOptions = connection_string(properties).parse()?;
    if properties.tls_ca.is_empty() {
        return Ok(options);
    }
    Ok(options.ssl_root_cert_from_pem(properties.tls_ca.clone().into_bytes()))
}

#[cfg(all(test, feature = "postgres"))]
mod tests {
    use super::*;
    use sqlx::postgres::PgSslMode;

    fn properties(tls_mode: &str, tls_ca: &str) -> PostgresProperties {
        PostgresProperties {
            host: "db.example.com".into(),
            port: 5432,
            database: "d".into(),
            username: "u".into(),
            password: "p".into(),
            tls_mode: tls_mode.into(),
            tls_ca: tls_ca.into(),
            ..Default::default()
        }
    }

    #[test]
    fn a_url_keeps_its_own_sslmode_and_options() {
        let options = connect_options(&PostgresProperties {
            url: "postgres://app:pw@ep-cool.neon.tech/orders?sslmode=require&options=endpoint%3Dep-cool".into(),
            ..Default::default()
        })
        .expect("options");
        assert_eq!(options.get_host(), "ep-cool.neon.tech");
        assert!(matches!(options.get_ssl_mode(), PgSslMode::Require));
        assert!(options
            .get_options()
            .is_some_and(|o| o.contains("endpoint")));
    }

    #[test]
    fn verify_full_trusts_the_records_ca() {
        let ca = "-----BEGIN CERTIFICATE-----\nMIIB\n-----END CERTIFICATE-----\n";
        let options = connect_options(&properties("verify-full", ca)).expect("options");
        assert!(matches!(options.get_ssl_mode(), PgSslMode::VerifyFull));
        assert!(format!("{options:?}").contains("ssl_root_cert: Some(Inline("));
    }

    #[test]
    fn no_mode_leaves_the_drivers_default() {
        let options = connect_options(&properties("", "")).expect("options");
        assert!(matches!(options.get_ssl_mode(), PgSslMode::Prefer));
    }
}
